### ISSUE TITLE
Multi gpu

### ISSUE THREAD

**Issue by mini-Shark** [MIC-DKFZ/nnUNet#82]
Hi thanks for your great work, i want to know does the nnUNet support multi-gpu training ? thanks :)

**Comment by FabianIsensee** [comment]
Hi,
right now it does not support it. I have an update coming where multi GPU will be supported.
Best,
Fabian

**Comment by mini-Shark** [comment]
Thanks for your reply, could I use `DistributedDataParallel`  to wrap model again for nnUNet to support multi GPU ?  Or could  you give me some advices to make nnUNet support multi GPU

**Comment by zhuczhuc** [comment]
I modified a version can run nnunet on multi GPU, on gitee https://gitee.com/IF10/nnUNet.

**Comment by FabianIsensee** [comment]
@zhuczhuc this may break some things (like the batch dice). And it will also put too much strain on GPU0. The load is not going to be well balanced and you will have a lot of GPU-GPU communication.

Same with DistributedDataParallel: You will run into problems with batch dice

**Comment by tiulpin** [comment]
As a workaround, you can run the training process with a variable `CUDA_VISIBLE_DEVICES` set to the wanted GPU number. For example, when I want to train the third fold on the third GPU, I run:
`CUDA_VISIBLE_DEVICES=3 python run/run_training.py 3d_fullres nnUNetTrainer TaskXX_MY_DATASET 3 --ndet`

**Comment by mini-Shark** [comment]
> As a workaround: you can run the training process with a variable `CUDA_VISIBLE_DEVICES` set to the wanted GPU number. For example, when I want to train the third fold on the third GPU, I run:
> 
> ```
> CUDA_VISIBLE_DEVICES=3 python run/run_training.py 3d_fullres nnUNetTrainer TaskXX_MY_DATASET 3 --ndet
> ```

hhaha Thanks

**Comment by zhuczhuc** [comment]
> @zhuczhuc this may break some things (like the batch dice). And it will also put too much strain on GPU0. The load is not going to be well balanced and you will have a lot of GPU-GPU communication.
> 
> Same with DistributedDataParallel: You will run into problems with batch dice

Yes, DataParallel make the load on GPU-0 is more than other 3, but training time one epoch doesn't change too much ( 10 % slower, I think) .And I use the same patch size, but the batch size is 2 on single GPU and 4 on 4 GPUs. Of course, I doesn't use SSD, maybe that's primary cause. 

I neglect the batch dice problems. Thanks for your mention! I will check it.

**Comment by FabianIsensee** [comment]
> As a workaround: you can run the training process with a variable CUDA_VISIBLE_DEVICES set to the wanted GPU number. For example, when I want to train the third fold on the third GPU, I run:
> 
> CUDA_VISIBLE_DEVICES=3 python run/run_training.py 3d_fullres nnUNetTrainer TaskXX_

Is that what you needed @mini-Shark ? If so then I did not understand your question correctly, sorry for that. Yes, CUDA_VISIBLE_DEVICES is the correct way to handle this!

> Yes, DataParallel make the load on GPU-0 is more than other 3, but training time one epoch doesn't change too much ( 10 % slower, I think) .And I use the same patch size, but the batch size is 2 on single GPU and 4 on 4 GPUs. Of course, I doesn't use SSD, maybe that's primary cause.

So you are using batch size 1 per gpu? This may not be an ideal usage of resources, because your are effectively trading a 2x increase in batch size with a 4x increase in compute resources. My DP implementation scales almost linearly, but I cannot share it right now. Please be patient :-)
(The trick is not to copy the softmax probabilities between GPUs)

**Comment by mini-Shark** [comment]
@FabianIsensee  I want use `DistributedDataParallel` to do multi gpu training for much more patch size  and batch size. thanks for your trick, I will try to implement it.

**Comment by FabianIsensee** [comment]
DistributedDataParallel is much more complicated if you want to implement it correctly (with batch dice). I have an implementation of that as well, but again - patience (sorry :-) )

**Comment by mini-Shark** [comment]
When are you going to release it (just curious) ? :eyes:

**Comment by FabianIsensee** [comment]
1-2 months

**Comment by wukaiyeah** [comment]
Hi,
If my computer-server doesn't have GPU, can I run nnUnet successfully?

**Comment by FabianIsensee** [comment]
No, nnU-Net does not support CPU. It would run way too slow on CPU anyways

**Comment by Yuhaooooo** [comment]
Hi FabianIsensee, any update on this? Thanks

**Comment by FabianIsensee** [comment]
Hi,
soonish :-D I am currently reworking my code and will do an update in a couple of weeks

**Comment by MengzhangLI** [comment]
Guten Morgen, Fabian. 
Will you upgrade nnUNet (using Multi-GPUs) in the next few weeks?
Looking forward to it very much. ;)

**Comment by FabianIsensee** [comment]
Most likely :-) Apologies things are crazy at the moment and it is hard to estimate. The code is ready (mostly)

**Comment by FabianIsensee** [comment]
Here you go. I just updated everything:
https://github.com/MIC-DKFZ/nnUNet#multi-gpu-training

**Comment by MengzhangLI** [comment]
> Here you go. I just updated everything:
> https://github.com/MIC-DKFZ/nnUNet#multi-gpu-training

Wow, it's amazing! Seems like this upgrade is very abundant.
Thank you very much. ;)

P.S. Take care in Europe. COVID-19 is terrible. Wish you all the best.

**Comment by mini-Shark** [comment]
> Here you go. I just updated everything:
> https://github.com/MIC-DKFZ/nnUNet#multi-gpu-training

Thanks for your great work

**Comment by chobe111** [comment]
> Here you go. I just updated everything:
> https://github.com/MIC-DKFZ/nnUNet#multi-gpu-training

Thank you so much

**Comment by FabianIsensee** [comment]
> > Here you go. I just updated everything:
> > https://github.com/MIC-DKFZ/nnUNet#multi-gpu-training
> 
> Wow, it's amazing! Seems like this upgrade is very abundant.
> Thank you very much. ;)
> 
> P.S. Take care in Europe. COVID-19 is terrible. Wish you all the best.

Thanks! We are all working from home right now :-) I hope it isn't too terrible where you live.
Best,
Fabian
